pennylane/transforms/decorators.py

```diff
@@ -0,0 +1,442 @@
+            qml.RX(0.1, wires=0)
+
+            # apply the transform to the ansatz
+            my_transform(*transform_weights)(ansatz)(param)
+
+            return qml.expval(qml.PauliZ(1))
+
+    We can print this QNode to show that the qfunc transform is taking place:
+
+    >>> x = np.array(0.5, requires_grad=True)
+    >>> y = np.array([0.1, 0.2], requires_grad=True)
+    >>> print(qml.draw(circuit)(x, y))
+     0: ──H────────────────────╭Z──┤
+     1: ──RX(0.05)──RY(0.141)──╰C──┤ ⟨Z⟩
+
+    Evaluating the QNode, as well as the derivative, with respect to the gate
+    parameter *and* the transform weights:
+
+    >>> circuit(x, y)
+    0.9887793925354269
+    >>> qml.grad(circuit)(x, y)
+    (array(-0.02485651), array([-0.02474011, -0.09954244]))
+    """
+
+    if not isinstance(tape_transform, single_tape_transform):
+        tape_transform = single_tape_transform(tape_transform)
+
+    sig = inspect.signature(tape_transform)
+    params = sig.parameters
+
+    if len(params) > 1:
+
+        @functools.wraps(tape_transform)
+        def make_qfunc_transform(*targs, **tkwargs):
+            def wrapper(fn):
+                @functools.wraps(fn)
+                def internal_wrapper(*args, **kwargs):
+                    tape = qml.transforms.make_tape(fn)(*args, **kwargs)
+                    tape = tape_transform(tape, *targs, **tkwargs)
+                    return tape.measurements
+
+                return internal_wrapper
+
+            return wrapper
+
+    elif len(params) == 1:
+
+        @functools.wraps(tape_transform)
+        def make_qfunc_transform(fn):
+            @functools.wraps(fn)
+            def internal_wrapper(*args, **kwargs):
+                tape = make_tape(fn)(*args, **kwargs)
+                tape = tape_transform(tape)
+                return tape.measurements
+
+            return internal_wrapper
+
+    make_qfunc_transform.tape_fn = tape_transform
+    return make_qfunc_transform
+
+
+def qnode_transform(qnode_transform_fn):
+    """Register a new QNode transform.
+
+    Args:
+        qnode_transform_fn (QNode transform): the QNode transform function
+            to register.
+
+            Allowed QNode transforms must be functions of the following form:
+
+            .. code-block:: python
+
+                def qnode_transform(qnode, *args, **kwargs):
+                    ...
+                    return tapes, processing_fn
+
+            That is, the first argument must be the input QNode to transform,
+            and the function must return a tuple ``(list, function)`` containing:
+
+            * A list of new tapes to execute, and
+            * A processing function with signature ``processing_fn(List[float])``
+              which is applied to the flat list of results from the executed tapes.
+
+            If ``tapes`` is empty, then it is assumed no quantum evaluations
+            are required, and ``processing_fn`` will be passed an empty list.

```

Comment:
I can't find the part of the code that handles this; in all cases there is the line 
```
res = [t.execute(device=qnode.device) for t in tapes]
```
which would execute the tapes, and fail is they are `None`.

Reply:
Oh, maybe I phrased it badly. Rather than `tapes` being `None`, I meant more that an empty list `tapes = []` would result in no quantum evaluations.

E.g., this allows you to do to create a QNode transform that performs no evaluations, such as recreating `qml.draw()`:

```python
@qml.qnode_transform
def draw(qnode):
    """Expand the QNode before drawing it"""
    tape = qnode.qtape.expand(depth=2)
    return [], lambda res: "My expanded circuit drawing!!!\n" + tape.draw()

dev = qml.device("default.qubit", wires=3)

@qml.qnode(dev)
def circuit(x):
    qml.RX(x, wires=0)
    qml.CNOT(wires=[0, 1])
    qml.CNOT(wires=[1, 2])
    qml.CNOT(wires=[2, 0])
    qml.Hadamard(wires=2)
    return qml.expval(qml.PauliZ(0))
```

```pycon
>>> print(draw(circuit)(0.5))
My expanded circuit drawing!!!
 0: ──RX(0.5)──╭C──────╭X────────────────────────────────┤ ⟨Z⟩
 1: ───────────╰X──╭C──│─────────────────────────────────┤
 2: ───────────────╰X──╰C──RZ(1.57)──RX(1.57)──RZ(1.57)──┤
```

Reply:
Although, this does raise the following question for me: is it nice that the tape evaluation is abstracted away? In most cases yes, unless you want fine-grained control over the execution.

E.g., say you have a QNode transform that creates multiple tapes executed on _different_ devices?

```python
@qml.qnode_transform
def complex_transform(qnode, params, devices):
    # this transform generates tapes, with each tape executed on a device from devices
    tapes = [tape_transform(qnode.qtape, p) for p in params]
    return tapes, lambda res: np.sum(np.sqrt(res))
```

We could expand the API so that the device list is 'known' to the decorator in two ways:

1. The decorator could simply look for a particular 'reserved' argument in the signature, such as `devices`. TensorFlow does something similar here.
2. Alternatively, we could add it to the return statement (`return tapes, fn, device_list`).

I tend to prefer option 1.

Reply:
I'm a fan of option 1 as well, it keeps the return statement fixed at `tapes, fn`.

Reply:
I don't see in your example how the device plays a role, though, since it is just the list of tapes that gets returned, without executing them. (I think I am still not fully understanding this, please let me know if I'm missing something)